### PR TITLE
algolia search

### PR DESCRIPTION
**Pull request by kailash-choudhary-thoughtspot**
replacing flex-search with Algolia search
1) search results will be sorted based on relevance instead of alphabetical order.
2) replaced keyword and page title for search result description with the section name and snippet of result(15 words)
3) directly navigate to the section for gatsby docs.
4) added highlight for search keyword in search results.
5) added search support for typedoc files.

### CODE REVIEW

File: docs/src/pages/index.tsx [hunk 1]
```diff
@@ -187,31 +187,55 @@ const IndexPage = ({ location }) => {
...
-    }, []);
+    const searchClient = React.useMemo(
+        () =>
+          algoliasearch(
+            process.env.GATSBY_ALGOLIA_APP_ID,
+            process.env.GATSBY_ALGOLIA_SEARCH_KEY
+          ),
+        []
+      );
+    const gatsbyIndex = searchClient.initIndex('Pages');
```
Comment (ashubham) [inline docs/src/pages/index.tsx:207]: can we rename `gatsbyIndex` to something related to search ?

File: docs/src/utils/algolia-queries.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+      return [
+        ...data.allAsciidoc.edges
+            .filter(
+                (edge) =>
+                    edge.node.pageAttributes.pageid &&
+                    edge.node.pageAttributes.pageid !== 'nav',
+            )
+            .reduce((acc,edge) => {
+                const pageid = edge.node.pageAttributes.pageid;
+                const newDiv = new JSDOM(`<div>${edge.node.html}</div>`).window.document;
```
Comment (ashubham) [inline docs/src/utils/algolia-queries.js:78]: Do we need to do `.window.document` ? DOM element should have `querySelector` method ?

File: docs/src/utils/algolia-queries.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+                  const sectTitle = sect.querySelector('h2').innerHTML;
+                  return {
+                    objectID: edge.node.id + sectId,
+                    sectionId: sectId,
+                    sectionTitle: sectTitle,
+                    body: sect && getTextFromHtml(sect.innerHTML),
+                    pageid,
+                    type: 'ASCII',
+                    title: edge.node.document.title,
+                  };
```
Comment (ashubham) [inline docs/src/utils/algolia-queries.js:100]: This code seems to be repeated multiple times, can we extract this into a method and call that with a `id`, `title` etc ?

**Review by ashubham — APPROVED** [review]

File: docs/src/utils/algolia-queries.js (hunk 2 above)
Comment (kailash-choudhary-thoughtspot) [reply docs/src/utils/algolia-queries.js:78]: with the latest JSDOM we need to do window.document. https://www.npmjs.com/package/jsdom

File: docs/src/pages/index.tsx (hunk 1 above)
Comment (kailash-choudhary-thoughtspot) [reply docs/src/pages/index.tsx:207]: ack

**Review by akashmohapatra — CHANGES_REQUESTED** [review]
@kailash-choudhary-thoughtspot We will need unit tests for the new code as well. The coverage of the docs folder contributes to the overall repo coverage

**Review by akashmohapatra — APPROVED** [review]
LGTM

**Pull request merged**